docs/internals/optimizer.rst

```diff
@@ -138,4 +184,669 @@ So the call to function ``f`` is inlined and the original definition of ``f`` ca
+ - Expression Simplifier
+ - Redundant Assign Eliminator
+ - Full Function Inliner
+
+Selecting optimizations
+~~~~~~~~~~~~~~~~~~~~~~~
+
+By default the optimizer applies its predefined sequence of optimization steps to
+the generated assembly. You can override this sequence and supply your own using
+the `--yul-optimizations` option:
+
+``` bash
+solc --optimize --ir-optimized --yul-optimizations 'dhfoD[xarrscLMcCTU]uljmul'
+```
+
+Available abbreviations are listed in the `Yul optimizer docs </docs/yul.rst#optimization-step-sequence>`_.
+
+Preprocessing
+~~~~~~~~~~~~~
+
+The preprocessing components perform transformations to get the program
+into a certain normal form that is easier to work with. This normal
+form is kept during the rest of the optimization process.
+
+Disambiguator
+~~~~~~~~~~~~~
+
+The disambiguator takes an AST and returns a fresh copy where all identifiers have
+names unique to the input AST. This is a prerequisite for all other optimizer stages.
+One of the benefits is that identifier lookup does not need to take scopes into account
+and we can basically ignore the result of the analysis phase.
+
+All subsequent stages have the property that all names stay unique. This means if
+a new identifier needs to be introduced, a new unique name is generated.
+
+Function Hoister
+~~~~~~~~~~~~~~~~
+
+The function hoister moves all function definitions to the end of the topmost block. This is
+a semantically equivalent transformation as long as it is performed after the
+disambiguation stage. The reason is that moving a definition to a higher-level block cannot decrease
+its visibility and it is impossible to reference variables defined in a different function.
+
+The benefit of this stage is that function definitions can be looked up more easily
+and functions can be optimized in isolation without having to traverse the AST.
+
+Function Grouper
+~~~~~~~~~~~~~~~~
+
+The function grouper has to be applied after the disambiguator and the function hoister.
+Its effect is that all topmost elements that are not function definitions are moved
+into a single block which is the first statement of the root block.
+
+After this step, a program has the following normal form:
+
+	{ I F... }
+
+Where I is a (potentially empty) block that does not contain any function definitions (not even recursively)
+and F is a list of function definitions such that no function contains a function definition.
+
+The benefit of this stage is that we always know where the list of function begins.
+
+### For Loop Condition Into Body
+
+This transformation moves the iteration condition of a for-loop into loop body.
+We need this transformation because [expression splitter](#expression-splitter) won't
+apply to iteration condition expressions (the `C` in the following example).
+
+    for { Init... } C { Post... } {
+        Body...
+    }
+
+is transformed to
+
+    for { Init... } 1 { Post... } {
+        if iszero(C) { break }
+        Body...
+    }
+
```

Comment:
Could also add something like "This transformation can also be useful when paired with Loop Invariant Code motion. Because invariants in the loop-invariant conditions can then be taken outside the loop."

Reply:
Added.